File: apps/desktop/src/renderer/services/workspaces.ts
import { BASE_URL, api } from "@/services/api";
import {
  type FilePreview,
  type WorkspaceEditDoc,
  type WorkspaceFile,
  type WorkspaceWriteOutcome,
  authedFetch,
  decodePreviewResponse,
  encodePath,
  saveBlob,
} from "@/services/workspaceHttp";
import type { components } from "@/types/api.generated";

type Schemas = components["schemas"];

/**
 * The first-class workspace REST client, addressed by **workspace id** (文件中枢
 * 统一 Step 1/2): `ws_id = "conv:<conversationId>"` (primary). This backs the file hub,
 * which browses *conversation scratch spaces* — distinct from `services/workspace` (the
 * per-conversation alias kept for the chat panel). Both hit the same server service
 * layer; only the addressing differs. File/CRUD here are valid for **cloud**
 * workspaces — local ones are reached over desktop IPC (the server returns 409),
 * so the hub picks `LocalRootSource` for those (§五).
 */

export interface WorkspaceInfo {
  wsId: string;
  name: string;
  location: "cloud" | "local";
  /** The bound desktop root id when local; null when cloud. */
  rootId: string | null;
  /** Sub-path within the bound local root (工作区对称化 D1a); "" = the root itself
   * (an explicitly-added local project) or cloud. A non-empty segment marks a
   * per-conversation workspace lazily promoted under a shared container root —
   * the hub scopes its browse ops to this subtree. */
  subpath: string;
  hasFiles: boolean;
}

/** Enumerate the user's workspaces for the hub rail. */
export async function listWorkspaces(): Promise<WorkspaceInfo[]> {
  const res = await api.get<Schemas["WorkspaceListResponse"]>("/v1/workspaces");
  return res.data.map((w) => ({
    wsId: w.ws_id,
    name: w.name,
    location: w.location,
    rootId: w.root_id ?? null,
    subpath: w.subpath ?? "",
    hasFiles: w.has_files,
  }));
}

const wsPath = (wsId: string): string =>
  `/v1/workspaces/${encodeURIComponent(wsId)}`;
const wsUrl = (wsId: string): string => `${BASE_URL}${wsPath(wsId)}`;

/** List a workspace's entries (recursive POSIX paths). */
export async function wsListFiles(
  wsId: string,
  recursive = true,
): Promise<WorkspaceFile[]> {
  const res = await api.get<Schemas["WorkspaceFileListResponse"]>(
    `${wsPath(wsId)}/files?recursive=${recursive}`,
  );
  return res.data.map((e) => ({ path: e.path, isDir: e.is_dir }));
}

/**
 * Flat file-path list for @ mentions (文件中枢统一 F4). Files only, ignore-pruned,
 * capped server-side — the cloud counterpart to `fsApi.listFiles` over a local
 * root, so @ indexes cloud and local workspaces the same way. Cloud-only (the
 * server refuses local workspaces with 409). `truncated` is dropped here: the @
 * index is best-effort, matching how local indexing ignores its own cap.
 */
export async function wsListFileIndex(wsId: string): Promise<string[]> {
  const res = await api.get<Schemas["WorkspaceFileIndexResponse"]>(
    `${wsPath(wsId)}/file-index`,
  );
  return res.data;
}

/** Upload (create/overwrite) a workspace file from raw bytes. */
export async function wsUploadFile(
  wsId: string,
  path: string,
  body: Blob,
): Promise<void> {
  await authedFetch(`${wsUrl(wsId)}/files/${encodePath(path)}`, {
    method: "PUT",
    body,
  });
}

/** Export a workspace Markdown file to a sibling ``.docx`` (server converter). */
export async function wsExportMdToDocx(
  wsId: string,
  path: string,
): Promise<{ path: string; warnings: string[] }> {
  const res = await api.post<{
    path: string;
    source_path: string;
    size_bytes: number;
    warnings: string[];
  }>(`${wsPath(wsId)}/export-docx`, { path });
  return { path: res.path, warnings: res.warnings ?? [] };
}

/** Stateless Markdown → Word (local desktop path; images as base64). */
export async function convertMdToDocx(input: {
  markdown: string;
  images: Record<string, string | null>;
  sourceName: string;
}): Promise<{
  docxBase64: string;
  warnings: string[];
  suggestedFilename: string;
}> {
  const res = await api.post<{
    docx_base64: string;
    warnings: string[];
    suggested_filename: string;
  }>("/v1/workspaces/convert/md-to-docx", {
    markdown: input.markdown,
    images: input.images,
    source_name: input.sourceName,
  });
  return {
    docxBase64: res.docx_base64,
    warnings: res.warnings ?? [],
    suggestedFilename: res.suggested_filename,
  };
}

/** Delete a workspace file or directory (directories go recursively). */
export async function wsDeleteFile(wsId: string, path: string): Promise<void> {
  await api.delete(`${wsPath(wsId)}/files/${encodePath(path)}`);
}

/** Move/rename a workspace file or directory (`AlreadyExists` → 422). */
export async function wsMoveFile(
  wsId: string,
  src: string,
  dst: string,
): Promise<void> {
  await api.post(`${wsPath(wsId)}/move`, { src, dst });
}

/** Create a workspace directory (parents created; `AlreadyExists` → 422). */
export async function wsCreateDir(wsId: string, path: string): Promise<void> {
  await api.post(`${wsPath(wsId)}/dirs`, { path });
}

/** Download a workspace file and save it via the browser. */
export async function wsDownloadFile(
  wsId: string,
  path: string,
  filename: string,
): Promise<void> {
  const res = await authedFetch(`${wsUrl(wsId)}/files/${encodePath(path)}`);
  await saveBlob(await res.blob(), filename);
}

/** Read a workspace file for read-only in-panel preview. */
export async function wsReadFile(
  wsId: string,
  path: string,
): Promise<FilePreview> {
  const res = await authedFetch(`${wsUrl(wsId)}/files/${encodePath(path)}`);
  return decodePreviewResponse(res, { path });
}

/** Server snapshot payload (`/v1/workspaces/{ws_id}/snapshots`). */
type BackendSnapshot = Schemas["SnapshotSummary"];

export interface WorkspaceSnapshot {
  snapshotId: string;
  label: string | null;
  createdAt: string;
  sizeBytes: number;
}

const toSnapshot = (s: BackendSnapshot): WorkspaceSnapshot => ({
  snapshotId: s.snapshot_id,
  label: s.label,
  createdAt: s.created_at,
  sizeBytes: s.size_bytes,
});

/** Take a manual snapshot of a cloud workspace addressed by ws id. */
export async function wsCreateSnapshot(
  wsId: string,
  label?: string,
): Promise<WorkspaceSnapshot> {
  const res = await api.post<BackendSnapshot>(`${wsPath(wsId)}/snapshots`, {
    label: label?.trim() || null,
  } satisfies Schemas["CreateSnapshotRequest"]);
  return toSnapshot(res);
}

/** blob → base64（分块，避免大文件撑爆调用栈）。 */
async function blobToBase64(blob: Blob): Promise<string> {
  const bytes = new Uint8Array(await blob.arrayBuffer());
  let binary = "";
  const chunk = 0x8000;
  for (let i = 0; i < bytes.length; i += chunk) {
    binary += String.fromCharCode(...bytes.subarray(i, i + chunk));
  }
  return btoa(binary);
}

/**
 * 「在浏览器打开」文件中枢云端工作区 HTML：ws 快照 → zip → 主进程解压临时目录 →
 * 系统默认浏览器。Shared spaces refuse snapshots (v1) → caller must not hang this
 * for `shared:` ws ids. Desktop-only (`previewArchive`).
 */
export async function openCloudWorkspaceInBrowser(
  wsId: string,
  htmlPath: string,
): Promise<void> {
  const preview = window.fsApi?.previewArchive;
  if (!preview) throw new Error("此环境不支持在浏览器打开");
  const snap = await wsCreateSnapshot(wsId, "浏览器预览");
  const res = await authedFetch(
    `${wsUrl(wsId)}/snapshots/${encodeURIComponent(snap.snapshotId)}/download`,
  );
  const archiveBase64 = await blobToBase64(await res.blob());
  const result = await preview(archiveBase64, htmlPath);
  if (!result.ok) throw new Error(result.message);
}

/** Read a cloud-workspace file for **editing** (full text + mtime CAS baseline). */
export async function wsReadFileForEdit(
  wsId: string,
  path: string,
): Promise<WorkspaceEditDoc> {
  const res = await api.get<Schemas["WorkspaceEditDoc"]>(
    `${wsPath(wsId)}/edit/${encodePath(path)}`,
  );
  return { text: res.text, mtimeMs: res.mtime_ms, eol: res.eol };
}

/** Conditionally write editor text back (mtime CAS); conflict carries disk mtime. */
export async function wsWriteFileText(
  wsId: string,
  path: string,
  input: { content: string; eol: "lf" | "crlf"; baselineMtimeMs: number },
): Promise<WorkspaceWriteOutcome> {
  const res = await api.put<Schemas["WorkspaceWriteResult"]>(
    `${wsPath(wsId)}/edit/${encodePath(path)}`,
    {
      content: input.content,
      eol: input.eol,
      baseline_mtime_ms: input.baselineMtimeMs,
    } satisfies Schemas["WorkspaceWriteRequest"],
  );
  return { ok: res.ok, mtimeMs: res.mtime_ms, conflict: res.conflict };
}

/** Shallow-clone an http(s) repo into a cloud workspace (G3). */
export async function wsCloneRepo(
  wsId: string,
  input: { repoUrl: string; dest?: string | null },
): Promise<string> {
  const res = await api.post<Schemas["CloneRepoResponse"]>(
    `${wsPath(wsId)}/clone`,
    {
      repo_url: input.repoUrl,
      dest: input.dest ?? null,
    } satisfies Schemas["CloneRepoRequest"],
  );
  return res.path;
}

File: apps/server/agentcore/api/schemas/workspaces.py
"""Workspace schemas: local-mode binding, snapshots, and workspace file ops."""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

# --- Workspace local-mode binding (双模式工作区 §七) ---


class BindLocalWorkspaceRequest(BaseModel):
    """Bind a 裸聊's scratch workspace to a desktop FS root (switch to local mode).

    ``root_id`` is the desktop-minted handle for an authorized local directory
    (from the desktop ``addRoot`` flow). Only ungrouped conversations may rebind —
    project chats inherit an immutable project binding (``PUT`` returns 409).
    """

    root_id: str = Field(..., min_length=1, max_length=200)


class WorkspaceBindingResponse(BaseModel):
    """A conversation's resolved workspace mode + where its binding lives.

    ``mode`` / ``root_id`` follow the turn-routing口径 (``resolve_local_binding``):
    an explicit ``local_root_id`` OR the desktop's ``local_container_root_id`` default
    both count as local. ``source`` distinguishes how the effective bind was chosen
    so the client can present「显式绑定」vs「容器默认」without guessing.
    """

    mode: Literal["local", "cloud"]
    # Which record carries the binding: the shared folder, or the conversation.
    scope: Literal["folder", "conversation"]
    # The bound desktop root id when local; None when cloud.
    root_id: str | None = None
    # How the effective local root was chosen. Absent/None when cloud.
    source: Literal["explicit", "container"] | None = None


# --- W3 session read-only external directory grants ---


class GrantExternalReadonlyRequest(BaseModel):
    """Register a session-scoped desktop root for this conversation.

    Does **not** change workspace binding. ``root_id`` is the desktop-minted handle;
    absolute paths never appear on the wire. ``mode`` is ``readonly`` (W3) or
    ``organize`` (P1: move/copy/mkdir/trash-delete).
    """

    root_id: str = Field(..., min_length=1, max_length=200)
    label: str = Field(..., min_length=1, max_length=200)
    alias_hint: str | None = Field(None, max_length=64)
    mode: Literal["readonly", "organize"] = "readonly"


class ExternalGrantItem(BaseModel):
    alias: str
    root_id: str
    label: str
    namespace: str  # ``external/<alias>``
    mode: Literal["readonly", "organize"] = "readonly"


class ExternalGrantListResponse(BaseModel):
    data: list[ExternalGrantItem]


class ExternalGrantResponse(BaseModel):
    grant: ExternalGrantItem


class WorkspaceSummary(BaseModel):
    """One addressable workspace in the file hub (文件中枢统一 Step 1).

    A folder project (``folder:<id>``) or an ungrouped conversation space
    (``conv:<id>``). ``location`` tells the hub how to reach its files: a cloud
    workspace via the ``/v1/workspaces/{ws_id}/files`` REST family; a local one
    over desktop IPC against ``root_id`` (its server-side dir is not the truth).
    """

    ws_id: str
    name: str
    location: Literal["cloud", "local"]
    # The bound desktop root id when local; None when cloud.
    root_id: str | None = None
    # Sub-path within ``root_id`` this workspace lives at (工作区对称化 D1a). Set for a
    # per-conversation local workspace lazily promoted under a shared container root;
    # None for cloud and for explicitly-added local projects bound at their root. The
    # desktop browses ``root_id`` + ``subpath`` so each sub-workspace shows only its
    # own files.
    subpath: str | None = None
    # Whether the space holds files. Folders always list (a project is a project);
    # ungrouped spaces list only when non-empty (F1) — so this is the filter that
    # let them in. Always True for local (the server can't see local files).
    has_files: bool


class WorkspaceListResponse(BaseModel):
    data: list[WorkspaceSummary]
    total: int


# --- Workspace snapshots ---


class CreateSnapshotRequest(BaseModel):
    """Take a manual snapshot of a conversation's workspace.

    A non-empty ``label`` marks a kept version (手动留版本) — a name the user
    pins — vs. the automatic post-turn backups.
    """

    label: str | None = Field(None, max_length=200)


class SnapshotSummary(BaseModel):
    """One persisted workspace snapshot (kept version or automatic backup)."""

    snapshot_id: str
    label: str | None
    created_at: datetime
    size_bytes: int

    model_config = {"from_attributes": True}


class SnapshotListResponse(BaseModel):
    data: list[SnapshotSummary]
    total: int


# --- Workspace files (bring files in / take results out) ---


class WorkspaceFileEntry(BaseModel):
    """One entry in a workspace listing — relative POSIX path + kind."""

    path: str
    is_dir: bool

    model_config = {"from_attributes": True}


class WorkspaceFileListResponse(BaseModel):
    data: list[WorkspaceFileEntry]
    total: int


class WorkspaceFileIndexResponse(BaseModel):
    """Flat file-path list for @ mentions (文件中枢统一 F4).

    Files only (no dirs), ignore-pruned, capped — ``truncated`` is True when the
    cap was hit. Mirrors the desktop ``fsApi.listFiles`` so cloud workspace files
    can feed the same @ index local roots already do.
    """

    data: list[str]
    total: int
    truncated: bool


class UploadFileResponse(BaseModel):
    """Result of a workspace file upload."""

    path: str
    size_bytes: int


class WorkspaceEditDoc(BaseModel):
    """Full text of a cloud workspace file for in-panel editing, plus CAS baseline.

    Unlike the preview download (truncated at the transfer cap), this returns the
    **whole** file so a save never drops the tail. ``mtime_ms`` is the write-time CAS
    baseline (compared on write); ``eol`` lets the editor restore the original line
    ending. Cloud files are server-stored UTF-8, so there is no encoding field.
    """

    text: str
    mtime_ms: int
    eol: Literal["lf", "crlf"]


class WorkspaceWriteRequest(BaseModel):
    """Conditional write of editor text to a cloud workspace file (mtime CAS).

    ``baseline_mtime_ms`` is the version the edit started from (``0`` = new file); a
    mismatch with the current disk mtime returns a conflict instead of clobbering an
    Agent's concurrent write. ``content`` uses ``\\n`` newlines; the server restores
    ``eol`` on write. Byte size is bounded in the route (same cap as upload).
    """

    content: str
    eol: Literal["lf", "crlf"] = "lf"
    baseline_mtime_ms: int = Field(0, ge=0)


class WorkspaceWriteResult(BaseModel):
    """Outcome of a conditional write.

    ``ok`` → ``mtime_ms`` is the new version (next baseline). On ``conflict`` →
    ``ok`` is False and ``mtime_ms`` is the **current disk** version, so the client
    can offer "overwrite anyway" by re-writing with it as the baseline.
    """

    ok: bool
    mtime_ms: int
    conflict: bool = False


class MoveFileRequest(BaseModel):
    """Move/rename a workspace file or directory (both workspace-relative)."""

    src: str = Field(..., min_length=1, max_length=1000)
    dst: str = Field(..., min_length=1, max_length=1000)


class CreateDirRequest(BaseModel):
    """Create a workspace directory (workspace-relative, parents created)."""

    path: str = Field(..., min_length=1, max_length=1000)


class CloneRepoRequest(BaseModel):
    """Clone a public git repository into the conversation's workspace."""

    repo_url: str = Field(..., min_length=1, max_length=2000)
    # Optional workspace-relative target dir; defaults to the repo name.
    dest: str | None = Field(None, max_length=500)


class CloneRepoResponse(BaseModel):
    """Result of a workspace clone — the relative dir the repo landed in."""

    path: str
